File: Jump_to_Algorithm2/week33/dayeongYoo/멀쩡한 사각형.java
// 가로 w, 세로 h인 직사각형, 1x1 정사각형으로 자를 예정
// 근데, 대각선 방향으로 잘라놈.
// 크기가 같은 직각삼각형 2개로 나눠짐
// 이때 사용할 수 있는 1x1 정사각형 개수?

import java.math.*;

class Solution {
    public long solution(int w, int h) { // 1억 이하 자연수
        long answer = 1;

        // 꼭짓점에서 만나는 점(사각형 중복됨) -> 최대공약수로 빼주기
        // 전체 사각형 개수
        long total = (long) w * h; // 1억이니까 long형변환

        // w,h 최대공약수
        long gcd = BigInteger.valueOf(w).gcd(BigInteger.valueOf(h)).longValue();

        // 사용할 수 없는 정사각형
        long unuse = w + h - gcd;

        answer = total - unuse;

        return answer;
    }
}

File: Jump_to_Algorithm2/week33/KyoungyeonJu/멀쩡한사각형.py
def solution(w,h):
    # 1억 100,000,000
    # 전체 w*h
    # 멀쩡한 사각형 (w-1)*(h-1)
    # 안 멀쩡한 사각형 w+h-1 = wh - (wh-w-h+1)
    def gcd(a, b):
        while b > 0:
            a, b = b, a % b
        return a

    gcd_num = gcd(w, h)
    nw, nh = w//gcd_num, h//gcd_num
    answer = w*h - (nw+nh-1)*gcd_num

    return answer